File: python/atlasent/trust_root.py
"""Hybrid trust-root bootstrap and snapshot management.

Loads the vendor snapshot at import time.  Optionally refreshes from
``https://keys.atlasent.io/.well-known/`` on a configurable interval
(default 4h, floor 5 min per ADR-005 D2).  Refresh failure is silent.

Snapshot expiry is fail-closed (ADR-005 D3): ``check_expiry()`` returns
``"expired"`` when ``valid_until`` passes, causing callers to raise
``BundleVerificationError(reason="trust_snapshot_expired")`` by default.
"""

from __future__ import annotations

import json
import logging
import threading
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Literal

logger = logging.getLogger(__name__)

# Half-life warning: emitted once per process
_half_life_warning_emitted = False
_expired_warning_emitted = False

REFRESH_INTERVAL_SECONDS_DEFAULT = 4 * 60 * 60  # 4 hours
REFRESH_INTERVAL_SECONDS_FLOOR = 5 * 60  # 5 minutes
KEYS_BASE_URL = "https://keys.atlasent.io/.well-known"

TrustKeyRole = Literal["R1_release", "R2_permit", "R3_audit", "R4_pack"]

_TRUST_ROOT_KEY_FIELDS = {
    "kid",
    "role",
    "kty",
    "alg",
    "x",
    "crv",
    "valid_from",
    "valid_until",
    "replaced_by",
    "revoked",
    "tenant",
}
_REVOCATION_ENTRY_FIELDS = {"kid", "revoked_at", "role", "reason"}


@dataclass
class TrustRootKey:
    kid: str
    role: str
    kty: str
    alg: str
    x: str | None = None
    crv: str | None = None
    valid_from: str | None = None
    valid_until: str | None = None
    replaced_by: str | None = None
    revoked: bool = False
    tenant: str | None = None


@dataclass
class TrustRootRevocationEntry:
    kid: str
    revoked_at: str
    role: str | None = None
    reason: str | None = None


@dataclass
class TrustRootSnapshot:
    valid_until: str
    issued_at: str
    keys: list[TrustRootKey] = field(default_factory=list)
    revoked_keys: list[TrustRootRevocationEntry] = field(default_factory=list)
    revoked_identities: list[dict[str, Any]] = field(default_factory=list)


def _make_trust_root_key(d: dict[str, Any]) -> TrustRootKey:
    return TrustRootKey(**{k: v for k, v in d.items() if k in _TRUST_ROOT_KEY_FIELDS})


def _make_revocation_entry(d: dict[str, Any]) -> TrustRootRevocationEntry:
    return TrustRootRevocationEntry(
        **{k: v for k, v in d.items() if k in _REVOCATION_ENTRY_FIELDS}
    )


class TrustRootManager:
    def __init__(
        self,
        initial_snapshot: TrustRootSnapshot,
        *,
        refresh_base_url: str = KEYS_BASE_URL,
        refresh_interval_seconds: int = REFRESH_INTERVAL_SECONDS_DEFAULT,
        disable_refresh: bool = False,
    ) -> None:
        self._snapshot = initial_snapshot
        self._lock = threading.RLock()
        self._refresh_base_url = refresh_base_url
        self._refresh_interval = max(
            refresh_interval_seconds, REFRESH_INTERVAL_SECONDS_FLOOR
        )
        self._timer: threading.Timer | None = None
        if not disable_refresh:
            self._schedule_refresh()

    def get_snapshot(self) -> TrustRootSnapshot:
        with self._lock:
            return self._snapshot

    def check_expiry(self) -> Literal["ok", "half_life", "expired"]:
        """Check whether the snapshot is expired; emit one-time warnings.

        Returns ``"ok"``, ``"half_life"``, or ``"expired"``.
        Emits a :mod:`logging` WARNING (via the ``atlasent`` logger) once
        per process for each of the half-life and expired conditions.
        """
        global _half_life_warning_emitted, _expired_warning_emitted

        snap = self.get_snapshot()
        from datetime import datetime, timezone

        now = datetime.now(timezone.utc)
        valid_until = datetime.fromisoformat(snap.valid_until.replace("Z", "+00:00"))
        issued_at = datetime.fromisoformat(snap.issued_at.replace("Z", "+00:00"))

        if now > valid_until:
            if not _expired_warning_emitted:
                _expired_warning_emitted = True
                days_ago = (now - valid_until).days
                logger.warning(
                    "[atlasent] Trust snapshot expired %d day(s) ago "
                    "(valid_until: %s). Update to a newer SDK build or "
                    "enable allow_expired_snapshot.",
                    days_ago,
                    snap.valid_until,
                )
            return "expired"
        window = (valid_until - issued_at).total_seconds()
        half_life = issued_at.timestamp() + window / 2
        if now.timestamp() > half_life:
            if not _half_life_warning_emitted:
                _half_life_warning_emitted = True
                days_left = (valid_until - now).days
                logger.warning(
                    "[atlasent] Trust snapshot expires in %d day(s) "
                    "(valid_until: %s). Plan an SDK update.",
                    days_left,
                    snap.valid_until,
                )
            return "half_life"
        return "ok"

    def lookup_key(self, kid: str) -> TrustRootKey | None:
        snap = self.get_snapshot()
        for k in snap.keys:
            if k.kid == kid:
                return k
        return None

    def is_revoked(self, kid: str) -> bool:
        snap = self.get_snapshot()
        return any(r.kid == kid for r in snap.revoked_keys)

    def replace_snapshot(self, next_snapshot: TrustRootSnapshot) -> None:
        with self._lock:
            self._snapshot = next_snapshot

    def stop_refresh(self) -> None:
        if self._timer is not None:
            self._timer.cancel()
            self._timer = None

    def _schedule_refresh(self) -> None:
        self._timer = threading.Timer(
            self._refresh_interval, self._do_refresh_and_reschedule
        )
        self._timer.daemon = True
        self._timer.start()

    def _do_refresh_and_reschedule(self) -> None:
        self._do_refresh()
        self._schedule_refresh()

    def _do_refresh(self) -> None:
        """Attempt a background refresh. All errors are caught and logged silently."""
        try:
            self._do_refresh_inner()
        except Exception as exc:
            logger.debug("trust-root refresh failed (silent): %s", exc)

    def _do_refresh_inner(self) -> None:
        import urllib.request

        base = self._refresh_base_url.rstrip("/")
        urls = {
            "index": f"{base}/atlasent-trust-root.json",
            "keys": f"{base}/atlasent-verifier-keys.json",
            "revocations": f"{base}/atlasent-revocations.json",
        }
        data: dict[str, Any] = {}
        for name, url in urls.items():
            with urllib.request.urlopen(url, timeout=10) as resp:  # noqa: S310
                data[name] = json.loads(resp.read())

        if not data.get("index", {}).get("valid_until"):
            return
        keys_data = data.get("keys", {}).get("keys", [])
        revoc_data = data.get("revocations", {})

        new_snap = TrustRootSnapshot(
            valid_until=data["index"]["valid_until"],
            issued_at=data["index"].get("issued_at", self._snapshot.issued_at),
            keys=[_make_trust_root_key(kd) for kd in keys_data],
            revoked_keys=[
                _make_revocation_entry(rd) for rd in revoc_data.get("revoked_keys", [])
            ],
            revoked_identities=revoc_data.get("revoked_identities", []),
        )
        self.replace_snapshot(new_snap)


# ─── Load the vendor snapshot ─────────────────────────────────────────────────

_VENDOR_DIR = Path(__file__).parent.parent.parent / "vendor" / "trust-root"


def _load_vendor_snapshot() -> TrustRootSnapshot:
    try:
        index = json.loads((_VENDOR_DIR / "atlasent-trust-root.json").read_text())
        verifier_keys_raw = json.loads(
            (_VENDOR_DIR / "atlasent-verifier-keys.json").read_text()
        )
        revocations_raw = json.loads(
            (_VENDOR_DIR / "atlasent-revocations.json").read_text()
        )
        keys = [_make_trust_root_key(kd) for kd in verifier_keys_raw.get("keys", [])]
        revoked_keys = [
            _make_revocation_entry(rd) for rd in revocations_raw.get("revoked_keys", [])
        ]
        return TrustRootSnapshot(
            valid_until=index["valid_until"],
            issued_at=index["issued_at"],
            keys=keys,
            revoked_keys=revoked_keys,
            revoked_identities=revocations_raw.get("revoked_identities", []),
        )
    except Exception:
        return TrustRootSnapshot(
            valid_until="2099-01-01T00:00:00Z",
            issued_at="2026-05-26T00:00:00Z",
            keys=[],
            revoked_keys=[],
            revoked_identities=[],
        )


_global_manager: TrustRootManager | None = None
_global_manager_lock = threading.Lock()


def get_global_trust_root_manager(
    *,
    disable_refresh: bool = False,
) -> TrustRootManager:
    global _global_manager
    with _global_manager_lock:
        if _global_manager is None:
            _global_manager = TrustRootManager(
                _load_vendor_snapshot(),
                disable_refresh=disable_refresh,
            )
    return _global_manager


def _set_global_trust_root_manager_for_tests(mgr: TrustRootManager | None) -> None:
    global _global_manager, _half_life_warning_emitted, _expired_warning_emitted
    _global_manager = mgr
    _half_life_warning_emitted = False
    _expired_warning_emitted = False

File: typescript/src/trustRoot.ts
/**
 * Hybrid trust-root bootstrap and snapshot management.
 *
 * Seeds synchronously from an embedded baseline snapshot (see
 * vendoredTrustRoot.generated.ts) — a plain object literal compiled
 * directly into this module, not read from disk at runtime. Optionally
 * refreshes from https://keys.atlasent.io/.well-known/ on a configurable
 * interval (default 4h, floor 5 min per ADR-005 D2). Refresh failure is
 * silent — falls back to the in-memory snapshot.
 *
 * Snapshot expiry (valid_until) is fail-closed per ADR-005 D3:
 * checkExpiry() emits a one-time console.warn at half-life, and again
 * on expiry. verifyAuditBundle throws BundleVerificationError when
 * expired (unless allowExpiredSnapshot=true is passed).
 *
 * CORRECTED: this module previously read vendor/trust-root/*.json via
 * fs.readFileSync(resolve(dirname(fileURLToPath(import.meta.url)), "..",
 * "..", ...)) at first use. Two independent defects made that path
 * unreachable in every real install: (1) package.json's `files` field
 * never listed `vendor`, so those JSON files were never included in the
 * published npm tarball; (2) even had they been included, tsup bundles
 * every entry into one flat file per format (dist/index.js), and going
 * up two directories from dist/ overshoots the package root by one level
 * — the math only ever "worked" by coincidence in this monorepo's dev
 * checkout, where vendor/ happens to sit two levels above typescript/src/
 * at the monorepo root. Confirmed by instantiating the real published
 * package (2.16.0) and reading its trust root: zero keys, zero
 * revocations, valid_until 2099 — the hardcoded empty fallback, always.
 * Every verifyBundle() call that didn't pass its own trustRoot explicitly
 * silently got a snapshot that can never detect a revoked key or a
 * role-mismatched key, contradicting this file's own ADR-005 D3/D4
 * fail-closed design. It was also a static node:fs/node:url/node:path
 * import, which broke bundling for any browser consumer regardless of
 * whether the file read itself would have succeeded. Embedding the
 * snapshot as a plain object literal (this module has zero imports of
 * its own) fixes both: no file I/O, no path guessing, nothing
 * Node-specific to bundle, and the data is present the instant the
 * module loads, in any environment.
 */

import { VENDORED_TRUST_ROOT_SNAPSHOT } from "./vendoredTrustRoot.generated.js";

// Types for the trust-root document shapes
export interface TrustRootKey {
  kid: string;
  role: "R1_release" | "R2_permit" | "R3_audit" | "R4_pack";
  kty: string;
  crv?: string;
  alg: string;
  x?: string;
  valid_from?: string | null;
  valid_until?: string | null;
  replaced_by?: string | null;
  revoked?: boolean;
  tenant?: string | null;
}

export interface TrustRootRevocationEntry {
  kid: string;
  role?: string;
  revoked_at: string;
  reason?: string;
}

export interface TrustRootSnapshot {
  /** ISO-8601 expiry of this snapshot; fail-closed when exceeded */
  valid_until: string;
  issued_at: string;
  keys: TrustRootKey[];
  revoked_keys: TrustRootRevocationEntry[];
  revoked_identities: Array<{ identity: string; revoked_at: string; reason?: string }>;
}

export interface TrustRootManagerOptions {
  /** Override the refresh URL (default: https://keys.atlasent.io/.well-known/) */
  refreshBaseUrl?: string;
  /** Refresh interval in ms. Default: 4h. Floor: 5 min. */
  refreshIntervalMs?: number;
  /** Disable automatic background refresh. */
  disableRefresh?: boolean;
  /** Custom fetch implementation (for tests). */
  fetch?: typeof fetch;
}

const REFRESH_INTERVAL_MS_DEFAULT = 4 * 60 * 60 * 1000; // 4 hours
const REFRESH_INTERVAL_MS_FLOOR = 5 * 60 * 1000; // 5 minutes
const KEYS_BASE_URL = "https://keys.atlasent.io/.well-known";

// Half-life and expiry warnings: emitted once per process (ADR-005 D3).
let _halfLifeWarningEmitted = false;
let _expiredWarningEmitted = false;

function _resetWarningFlags(): void {
  _halfLifeWarningEmitted = false;
  _expiredWarningEmitted = false;
}

export class TrustRootManager {
  private _snapshot: TrustRootSnapshot;
  private _refreshTimer: ReturnType<typeof setInterval> | null = null;
  private readonly _opts: Required<TrustRootManagerOptions>;

  constructor(
    initialSnapshot: TrustRootSnapshot,
    opts: TrustRootManagerOptions = {},
  ) {
    this._snapshot = initialSnapshot;
    const intervalMs = Math.max(
      opts.refreshIntervalMs ?? REFRESH_INTERVAL_MS_DEFAULT,
      REFRESH_INTERVAL_MS_FLOOR,
    );
    this._opts = {
      refreshBaseUrl: opts.refreshBaseUrl ?? KEYS_BASE_URL,
      refreshIntervalMs: intervalMs,
      disableRefresh: opts.disableRefresh ?? false,
      fetch:
        opts.fetch ??
        (typeof globalThis !== "undefined" && globalThis.fetch
          ? globalThis.fetch.bind(globalThis)
          : ((_url: string) =>
              Promise.reject(new Error("fetch not available"))) as typeof fetch),
    };
    if (!this._opts.disableRefresh) {
      this._scheduleRefresh();
    }
  }

  getSnapshot(): TrustRootSnapshot {
    return this._snapshot;
  }

  /**
   * Check whether the snapshot is expired, emit one-time warnings at
   * half-life and expiry.  Returns "ok" | "half_life" | "expired".
   *
   * Emits console.warn once per process at half-life (ADR-005 D3).
   * Emits console.warn once per process on expiry.
   */
  checkExpiry(): "ok" | "half_life" | "expired" {
    const snap = this._snapshot;
    const now = Date.now();
    const issuedAt = new Date(snap.issued_at).getTime();
    const validUntil = new Date(snap.valid_until).getTime();

    if (now > validUntil) {
      if (!_expiredWarningEmitted) {
        _expiredWarningEmitted = true;
        const daysAgo = Math.floor((now - validUntil) / (24 * 60 * 60 * 1000));
        // eslint-disable-next-line no-console
        console.warn(
          `[atlasent] Trust snapshot expired ${daysAgo} day(s) ago (valid_until: ${snap.valid_until}). ` +
            "Update to a newer SDK build or enable allowExpiredSnapshot.",
        );
      }
      return "expired";
    }
    const window = validUntil - issuedAt;
    const halfLife = issuedAt + window / 2;
    if (now > halfLife) {
      if (!_halfLifeWarningEmitted) {
        _halfLifeWarningEmitted = true;
        const daysLeft = Math.floor((validUntil - now) / (24 * 60 * 60 * 1000));
        // eslint-disable-next-line no-console
        console.warn(
          `[atlasent] Trust snapshot at half-life: expires in ${daysLeft} day(s) (valid_until: ${snap.valid_until}). ` +
            "Plan an SDK update.",
        );
      }
      return "half_life";
    }
    return "ok";
  }

  /** Look up a key entry by kid. Returns undefined if not found. */
  lookupKey(kid: string): TrustRootKey | undefined {
    return this._snapshot.keys.find((k) => k.kid === kid);
  }

  /** Returns true if the kid appears in revoked_keys. */
  isRevoked(kid: string): boolean {
    return this._snapshot.revoked_keys.some((r) => r.kid === kid);
  }

  /** Replace the snapshot (e.g. after a successful refresh). */
  replaceSnapshot(next: TrustRootSnapshot): void {
    this._snapshot = next;
  }

  stopRefresh(): void {
    if (this._refreshTimer !== null) {
      clearInterval(this._refreshTimer);
      this._refreshTimer = null;
    }
  }

  private _scheduleRefresh(): void {
    this._refreshTimer = setInterval(() => {
      void this._doRefresh();
    }, this._opts.refreshIntervalMs);
    // Don't hold the process open.
    if (
      this._refreshTimer &&
      typeof this._refreshTimer === "object" &&
      "unref" in this._refreshTimer
    ) {
      (this._refreshTimer as { unref(): void }).unref();
    }
  }

  private async _doRefresh(): Promise<void> {
    try {
      const base = this._opts.refreshBaseUrl.replace(/\/$/, "");
      const [keysRes, revocRes] = await Promise.all([
        this._opts.fetch(`${base}/atlasent-verifier-keys.json`),
        this._opts.fetch(`${base}/atlasent-revocations.json`),
      ]);
      const indexRes = await this._opts.fetch(`${base}/atlasent-trust-root.json`);

      if (!keysRes.ok || !revocRes.ok || !indexRes.ok) return;

      const [keys, revoc, index] = await Promise.all([
        keysRes.json() as Promise<{ keys: TrustRootKey[] }>,
        revocRes.json() as Promise<{
          revoked_keys: TrustRootRevocationEntry[];
          revoked_identities: unknown[];
        }>,
        indexRes.json() as Promise<{ valid_until: string; issued_at: string }>,
      ]);

      if (!index.valid_until || !Array.isArray(keys.keys)) return;

      this._snapshot = {
        valid_until: index.valid_until,
        issued_at: index.issued_at ?? this._snapshot.issued_at,
        keys: keys.keys,
        revoked_keys: revoc.revoked_keys ?? [],
        revoked_identities:
          (revoc.revoked_identities as Array<{
            identity: string;
            revoked_at: string;
          }>) ?? [],
      };
    } catch {
      // Refresh failure is silent — keep using the current snapshot.
    }
  }
}

// ─── Load the embedded (vendored) snapshot ───────────────────────────────────

function _loadVendorSnapshot(): TrustRootSnapshot {
  // The embedded constant is committed source (see vendoredTrustRoot.generated.ts),
  // always present at build time — there is no I/O and nothing to fall back from.
  return VENDORED_TRUST_ROOT_SNAPSHOT;
}

// Process-global manager — created lazily.
let _globalManager: TrustRootManager | null = null;

export function getGlobalTrustRootManager(
  opts?: TrustRootManagerOptions,
): TrustRootManager {
  if (!_globalManager) {
    _globalManager = new TrustRootManager(
      _loadVendorSnapshot(),
      opts ?? { disableRefresh: false },
    );
  }
  return _globalManager;
}

/** Replace the global manager (primarily for tests). */
export function __setGlobalTrustRootManagerForTests(
  mgr: TrustRootManager | null,
): void {
  _globalManager = mgr;
  _resetWarningFlags();
}

export { _resetWarningFlags as __resetWarningFlagsForTests };
